### PR TITLE
splitting sfs by mutation types

### PR DESCRIPTION
**Pull request by mufernando**
With popsim-consortium/stdpopsim#1152, now we store information about the DFEs and mutation types in the metadata of the tree sequence.

In this PR, I changed the way we split sites by class (neutral vs. non-neutral) using the metadata.

working on this with @izabelcavassim

### CODE REVIEW

File: benchmark/Snakefile [hunk 1]
```diff
@@ -108,33 +108,44 @@ rule simulation:
...
-            mut_type[j] = 3
-        else:
-            mut_type[j] = mt[0]
-
-    freqs = allele_counts(ts, [sample_sets])
+        site_class[j] = mut_types[mt[0]] if len(mt) == 1 else "double_hit" 
+    assert sum(site_class == None) == 0
+    print("Number of sites per class is:", flush=True)
+    unique, counts = numpy.unique(site_class, return_counts=True)
+    print(dict(zip(unique, counts)))
```
Comment (andrewkern) [inline benchmark/Snakefile:141]: so i think this code should be pushed to its own module in analysis, perhaps called `SFS_tools.py` and then imported to this workflow

File: benchmark/Snakefile [hunk 2]
```diff
@@ -288,8 +299,8 @@ rule generate_dadi_fs:
         else: samps = ts.samples()[:sizes[index]]
         mut_afs = generate_fs_from_ts(ts, samps)
 
-        nonneu_fs = dadi.Spectrum(mut_afs[:,0])
-        neu_fs = dadi.Spectrum(mut_afs[:,1])
+        nonneu_fs = dadi.Spectrum(mut_afs["non_neutral"])
```
Comment (andrewkern) [inline benchmark/Snakefile:302]: yep and this whole function too!

**Review by andrewkern — CHANGES_REQUESTED** [review]
I marked a few places in the PR that should be split out of the Snakefile and put into its own module

File: benchmark/Snakefile [hunk 3]
```diff
@@ -108,33 +108,44 @@ rule simulation:
...
                                     polarised=True, mode='site', strict=False)
 
-    mut_type = np.zeros(ts.num_sites)
+    # Mapping mutation type IDs to class of mutation (e.g., neutral, non-neutral)
+    mut_types = {}
+    for dfe in ts.metadata["stdpopsim"]["DFEs"]:
+        for mt in dfe["mutation_types"]:
+            mid = mt["slim_mutation_type_id"]
+            if not mid in mut_types:
+                mut_types[mid] = "neutral" if mt["is_neutral"] else "non_neutral"
```
Comment (petrelharp) [inline benchmark/Snakefile:126]:
```suggestion
            label = "neutral" if mt["is_neutral"] else "non_neutral"
            if mid in mut_types:
                assert mut_types[mid] == label
            else:
                mut_types[mid] = label
```

File: benchmark/Snakefile [hunk 4]
```diff
@@ -108,33 +108,44 @@ rule simulation:
...
             for md in m.metadata["mutation_list"]:
                 mt.append(md["mutation_type"])
-        if len(set(mt)) > 1:
-            mut_type[j] = 3
-        else:
-            mut_type[j] = mt[0]
-
-    freqs = allele_counts(ts, [sample_sets])
+        site_class[j] = mut_types[mt[0]] if len(mt) == 1 else "double_hit" 
+    assert sum(site_class == None) == 0
```
[the remark this answers is not in the document]
Comment (petrelharp) [reply benchmark/Snakefile:138]: oh wait, that maybe doesn't work with numpy arrays.

**Comment by petrelharp** [comment]
So - I am not actually sure what to do about the conflicts between this PR and #31???

**Comment by mufernando** [comment]
Let's wait for #31 to go in, then I can apply my changes to the new file @xin-huang created. The only thing that matters in this PR is the change in the way we split the SFS by mutation class (now using the metadata).

**Comment by izabelcavassim** [comment]
I will get my part of the analyses in once we get the #31 in.

**Comment by andrewkern** [comment]
okay so #31 is in.  @xin-huang split these functions into a new module called `ts2fs.py` and he split the snakefile into two now. I'm not sure if we need this PR anymore @mufernando ?

**Pull request closed without merging**